File: cmajor/parsing/Grammar.cpp
// =================================
// Copyright (c) 2019 Seppo Laakko
// Distributed under the MIT license
// =================================

#include <cmajor/parsing/Grammar.hpp>
#include <cmajor/parsing/Scope.hpp>
#include <cmajor/parsing/Rule.hpp>
#include <cmajor/parsing/Exception.hpp>
#include <cmajor/parsing/Visitor.hpp>
#include <cmajor/parsing/Composite.hpp>
#include <cmajor/parsing/XmlLog.hpp>
#include <cmajor/parsing/Linking.hpp>
#include <cmajor/parsing/ParsingDomain.hpp>
#include <cmajor/util/Unicode.hpp>

namespace cmajor { namespace parsing {

using namespace cmajor::util;
using namespace cmajor::unicode;

Grammar::Grammar(const std::u32string& name_, Scope* enclosingScope_) :
    ParsingObject(name_, enclosingScope_, ObjectKind::grammar), parsingDomain(new ParsingDomain()), ns(nullptr),
    linking(false), linked(false), contentParser(nullptr), startRule(nullptr), skipRule(nullptr), log(0), maxLogLineLength(80)
{
    RegisterParsingDomain(parsingDomain);
    SetScope(new Scope(Name(), EnclosingScope()));
}

Grammar::Grammar(const std::u32string& name_, Scope* enclosingScope_, ParsingDomain* parsingDomain_) : 
    ParsingObject(name_, enclosingScope_, ObjectKind::grammar), parsingDomain(parsingDomain_), ns(nullptr),
    linking(false), linked(false), contentParser(nullptr), startRule(nullptr), skipRule(nullptr), log(0), maxLogLineLength(80)
{
    SetScope(new Scope(Name(), EnclosingScope()));
}

void Grammar::AddRule(Rule* rule)
{
    Own(rule);
    rule->SetGrammar(this);
    rule->GetScope()->SetEnclosingScope(GetScope());
    rule->SetEnclosingScope(GetScope());
    rules.push_back(rule);
    try
    {
        GetScope()->Add(rule);
    }
    catch (std::exception& ex)
    {
        ThrowException(ex.what(), GetSpan());
    }
}

Rule* Grammar::GetRule(const std::u32string& ruleName) const
{
    ParsingObject* object = GetScope()->Get(ruleName, ObjectKind::rule | ObjectKind::ruleLink);
    if (!object)
    {
        ThrowException("rule '" + ToUtf8(ruleName) + "' not found in grammar '" + ToUtf8(FullName()) + "'", GetSpan());
    }
    if (object->IsRule())
    {
        return static_cast<Rule*>(object);
    }
    else if (object->IsRuleLink())
    {
        RuleLink* ruleLink = static_cast<RuleLink*>(object);
        Rule* rule = ruleLink->GetRule();
        if (rule)
        {
            return rule;
        }
        else
        {
            ThrowException("rule link '" + ToUtf8(ruleName) + "' not bound in grammar '" + ToUtf8(FullName()) + "'", GetSpan());
        }
    }
    else
    {
        ThrowException("'" + ToUtf8(ruleName) + "' is not a rule or rule link", GetSpan());
    }
    return nullptr;
}

void Grammar::Accept(Visitor& visitor)
{
    visitor.BeginVisit(*this);
    RuleLinkSetIt e = ruleLinks.cend();
    for (RuleLinkSetIt i = ruleLinks.cbegin(); i != e; ++i)
    {
        RuleLink* ruleLink = *i;
        ruleLink->Accept(visitor);
    }
    int n = int(rules.size());
    for (int i = 0; i < n; ++i)
    {
        Rule* rule = rules[i];
        rule->Accept(visitor);
    }
    visitor.EndVisit(*this);
}

void Grammar::Parse(const char32_t* start, const char32_t* end, int fileIndex, const std::string& fileName, ParsingData* parsingData)
{
    Scanner scanner(start, end, fileName, fileIndex, skipRule);
    scanner.SetParsingData(parsingData);
    std::unique_ptr<XmlLog> xmlLog;
    if (log)
    {
        xmlLog.reset(new XmlLog(*log, maxLogLineLength));
        scanner.SetLog(xmlLog.get());
        xmlLog->WriteBeginRule("parse");
        xmlLog->IncIndent();
    }
    ObjectStack stack;
    Match match = Parse(scanner, stack, parsingData);
    Span stop = scanner.GetSpan();
    if (log)
    {
        xmlLog->DecIndent();
        xmlLog->WriteEndRule("parse");
    }
    if (!match.Hit() || stop.Start() != int(end - start))
    {
        if (startRule)
        {
            throw ExpectationFailure(startRule->Info(), fileName, stop, start, end);
        }
        else
        {
            throw ParsingException("grammar '" + ToUtf8(FullName()) + "' has no start rule", fileName, stop, start, end);
        }
    }
}

Match Grammar::Parse(Scanner& scanner, ObjectStack& stack, ParsingData* parsingData)
{
    if (startRule)
    {
        if (!contentParser)
        {
            if (skipRule)
            {
                contentParser = new SequenceParser(new SequenceParser(new OptionalParser(skipRule), startRule), new OptionalParser(skipRule));
                Own(contentParser);
            }
            else
            {
                contentParser = startRule;
            }
        }
        return contentParser->Parse(scanner, stack, parsingData);
    }
    return Match::Nothing();
}

void Grammar::Parse(const char32_t* start, const char32_t* end, int fileIndex, const std::string& fileName)
{
    std::unique_ptr<ParsingData> parsingData(new ParsingData(parsingDomain->GetNumRules()));
    Parse(start, end, fileIndex, fileName, parsingData.get());
}

Match Grammar::Parse(Scanner& scanner, ObjectStack& stack)
{
    std::unique_ptr<ParsingData> parsingData(new ParsingData(parsingDomain->GetNumRules()));
    return Parse(scanner, stack, parsingData.get());
}

void Grammar::ResolveStartRule()
{
    if (startRuleName.empty())
    {
        if (!rules.empty())
        {
            startRule = rules.front();
        }
        else
        {
            ThrowException("cannot resolve start rule because grammar '" + ToUtf8(FullName()) + "' is empty", GetSpan());
        }
    }
    else
    {
        startRule = GetRule(startRuleName);
    }
}

void Grammar::ResolveSkipRule()
{
    if (!skipRuleName.empty())
    {
        skipRule = GetRule(skipRuleName);
    }
}

void Grammar::Link()
{
    if (!linked && !linking)
    {
        linking = true;
        GetReferencedGrammars();
        GrammarSetIt e = grammarReferences.end();
        for (GrammarSetIt i = grammarReferences.begin(); i != e; ++i)
        {
            Grammar* grammarReference = *i;
            grammarReference->Link();
        }
        LinkerVisitor linkerVisitor;
        Accept(linkerVisitor);
        linking = false;
        linked = true;
    }
}

void Grammar::AddGrammarReference(Grammar* grammarReference)
{
    Own(grammarReference);
    grammarReferences.insert(grammarReference);
}

void Grammar::AddRuleLink(RuleLink* ruleLink)
{
    Own(ruleLink);
    ruleLinks.insert(ruleLink);
}

} } // namespace cmajor::parsing

File: devenv/soul.parsing/Grammar.cs
/*========================================================================
    Copyright (c) 2011-2012 Seppo Laakko
    http://sourceforge.net/projects/soulparsing/
 
    Distributed under the GNU General Public License, version 3 (GPLv3).
    (See accompanying LICENSE.txt or http://www.gnu.org/licenses/gpl.html)

========================================================================*/

using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace soul.parsing
{
    public class LinkerVisitor : Visitor
    {
        public LinkerVisitor(Grammar grammar)
        {
            this.grammar = grammar;
        }
        public override void BeginVisit(RuleParser parser)
        {
            currentRule = parser;
            actionNumber = 0;
        }
        public override void EndVisit(RuleParser parser)
        {
            currentRule.Link();
        }
        public override void BeginVisit(ActionParser parser)
        {
            if (string.IsNullOrEmpty(parser.Name))
            {
                parser.Name = "A" + actionNumber.ToString();
                ++actionNumber;
            }
            currentRule.AddAction(parser);
        }
        public override void Visit(NonTerminalParser parser)
        {
            try
            {
                string ruleName = parser.RuleName;
                parser.Rule = grammar.GetRule(ruleName);
                if (parser.Specialized)
                {
                    currentRule.AddNonTerminal(parser);
                }
                if (parser.NumberOfArguments != parser.Rule.NumberOfParameters)
                {
                    throw new Exception("Rule '" + parser.RuleName + "' takes " + parser.Rule.NumberOfParameters.ToString() + " parameters (" +
                        parser.NumberOfArguments.ToString() + " arguments supplied)");
                }

            }
            catch (Exception ex)
            {
                if (parser.Position.Line != 0)
                {
                    throw new PositionException(ex.Message, parser.Position);
                }
                else
                {
                    throw ex;
                }
            }
        }
        public override void Visit(KeywordParser parser)
        {
            string continuationRuleName = parser.ContinuationRuleName;
            if (!string.IsNullOrEmpty(continuationRuleName))
            {
                parser.SetContinuationRule(grammar.GetRule(continuationRuleName));
            }
        }
        public override void Visit(KeywordListParser parser)
        {
            string selectorRuleName = parser.SelectorRuleName;
            parser.SetSelectorRule(grammar.GetRule(selectorRuleName));
        }
        public override void Visit(KeywordRuleListParser parser)
        {
            string selectorRuleName = parser.SelectorRuleName;
            parser.SetSelectorRule(grammar.GetRule(selectorRuleName));
            foreach (KeywordRule keywordRule in parser.KeywordRuleList)
            {
                parser.SetKeywordRule(keywordRule.Keyword, grammar.GetRule(keywordRule.RuleName));
            }
        }
        private Grammar grammar;
        private RuleParser currentRule;
        private int actionNumber;
    }

    public class GrammarReference
    {
        public GrammarReference(string grammarFullName, string assemblyFileName)
        {
            this.grammarFullName = grammarFullName;
            this.assemblyFileName = assemblyFileName;
        }
        public string GrammarFullName
        {
            get { return grammarFullName; }
        }
        public string AssemblyFileName
        {
            get { return assemblyFileName; }
        }
        private string grammarFullName;
        private string assemblyFileName;
    }

    public class Grammar : RuleParser
    {
        public Grammar(string name, string fullName)
            : this(name, fullName, new ParsingDomain())
        {
        }
        public Grammar(string name, string fullName, ParsingDomain parsingDomain)
            : base(name, fullName, false)
        {
            this.parsingDomain = parsingDomain;
            parsingDomain.AddGrammar(this);
            startRuleName = null;
            start = null;
            skipRuleName = null;
            skip = null;
            grammarReferences = new List<GrammarReference>();
            referencedGrammars = new List<Grammar>();
            rules = new List<RuleParser>();
            ruleScope = new Scope<RuleParser>(name + ".rules", "Rule");
            linking = false;
            linked = false;
            CreateRules();
            maxLogLineLength = 79;
        }
        public ParsingDomain ParsingDomain
        {
            get { return parsingDomain; }
        }
        public List<RuleParser> Rules
        {
            get { return rules; }
        }
        public Scope<RuleParser> RuleScope
        {
            get { return ruleScope; }
        }
        public string StartRuleName
        {
            get { return startRuleName; }
            set { startRuleName = value; }
        }
        public RuleParser Start
        {
            get { return start; }
            set { start = value; }
        }
        public string SkipRuleName
        {
            get { return skipRuleName; }
            set { skipRuleName = value; }
        }
        public RuleParser Skip
        {
            get { return skip; }
            set { skip = value; }
        }
        public string SynchronizationRuleName
        {
            get { return synchronizationRuleName; }
            set { synchronizationRuleName = value; }
        }
        public RuleParser SynchronizationRule
        {
            get { return synchronizationRule; }
            set { synchronizationRule = value; }
        }
        public TextWriter Log
        {
            get { return log; }
            set { log = value; }
        }
        public int LogLineLength
        {
            get { return maxLogLineLength; }
            set { maxLogLineLength = value; }
        }
        public List<GrammarReference> GrammarReferences
        {
            get { return grammarReferences; }
        }
        public List<Grammar> ReferencedGrammars
        {
            get { return referencedGrammars; }
        }
        public void AddGrammarReference(GrammarReference grammarReference)
        {
            grammarReferences.Add(grammarReference);
        }
        public virtual void CreateRules()
        {
        }
        public virtual void GetReferencedGrammars()
        {
        }
        public override List<AttrOrVariable> InheritedAttributes
        {
            get { return start != null ? start.InheritedAttributes : base.InheritedAttributes; }
        }
        public override string ValueTypeName
        {
            get { return start != null ? start.ValueTypeName : base.ValueTypeName; }
            set { if (start != null) start.ValueTypeName = value; else base.ValueTypeName = value; }
        }
        public override List<AttrOrVariable> LocalVariables
        {
            get { return start != null ? start.LocalVariables : base.LocalVariables; }
        }
        public void AddRule(RuleParser rule)
        {
            ruleScope.Add(rule);
            rules.Add(rule);
        }
        public RuleParser GetRule(string ruleName)
        {
            return ruleScope.Get(ruleName);
        }
        public override void Link()
        {
            if (!linked && !linking)
            {
                linking = true;
                GetReferencedGrammars();
                foreach (Grammar grammar in referencedGrammars)
                {
                    grammar.Link();
                }
                MergeRuleScopes();
                LinkerVisitor linkerVisitor = new LinkerVisitor(this);
                Accept(linkerVisitor);
                ResolveStart();
                ResolveSkip();
                ResolveSynchronizationRule();
                linking = false;
                linked = true;
            }
        }
        public override void ExpandCode()
        {
            foreach (RuleParser rule in rules)
            {
                rule.ExpandCode();
            }
        }
        private void ResolveStart()
        {
            if (string.IsNullOrEmpty(startRuleName))
            {
                foreach (RuleParser rule in rules)
                {
                    if (!rule.IsExternal)
                    {
                        start = rule;
                        return;
                    }
                }
            }
            else
            {
                start = ruleScope.Get(startRuleName);
            }
        }
        private void ResolveSkip()
        {
            if (!string.IsNullOrEmpty(skipRuleName))
            {
                skip = ruleScope.Get(skipRuleName);
            }
        }
        private void ResolveSynchronizationRule()
        {
            if (!string.IsNullOrEmpty(synchronizationRuleName))
            {
                synchronizationRule = ruleScope.Get(synchronizationRuleName);
            }
        }
        private void MergeRuleScopes()
        {
            foreach (Grammar grammar in referencedGrammars)
            {
                ruleScope.Merge(grammar.ruleScope);
            }
        }
        public void Parse(string content, int fileIndex, string fileName)
        {
            Scanner scanner = new Scanner(content, fileIndex, fileName, skip);
            if (log != null)
            {
                scanner.Log = new XmlLog(log, maxLogLineLength);
                scanner.Log.WriteBeginRule("parse");
            }
            Stack<object> stack = new Stack<object>();
            Match match = Parse(scanner, stack);
            Position stop = scanner.Position;
            if (log != null)
            {
                scanner.Log.WriteEndRule("parse");
            }
            if (!match.Hit || stop.Index != content.Length)
            {
                if (start != null)
                {
                    throw new ExpectationFailure(start.Info, content, scanner.Position, scanner.FileName);
                }
                else
                {
                    throw new ParsingException("grammar '" + Name + "' has no start rule", content, scanner.Position, scanner.FileName);
                }
            }
        }
        public override Match Parse(Scanner scanner, Stack<object> stack)
        {
            if (start != null)
            {
                if (content == null)
                {
                    if (skip != null)
                    {
                        content = new SequenceParser(
                            new SequenceParser(
                                new OptionalParser(skip),
                                start),
                            new OptionalParser(skip));
                    }
                    else
                    {
                        content = start;
                    }
                }
                return content.Parse(scanner, stack);
            }
            return Match.Nothing();
        }
        public override void Accept(Visitor visitor)
        {
            visitor.BeginVisit(this);
            foreach (RuleParser rule in rules)
            {
                rule.Accept(visitor);
            }
            visitor.EndVisit(this);
        }
        private ParsingDomain parsingDomain;
        private string startRuleName;
        private RuleParser start;
        private string skipRuleName;
        private RuleParser skip;
        private string synchronizationRuleName;
        private RuleParser synchronizationRule;
        private List<GrammarReference> grammarReferences;
        private List<Grammar> referencedGrammars;
        private List<RuleParser> rules;
        private Scope<RuleParser> ruleScope;
        private bool linked;
        private bool linking;
        private TextWriter log;
        private int maxLogLineLength;
        private Parser content;
    }
}
